src/core/middlewares/sizequota/handler.go

```diff
@@ -0,0 +1,290 @@
+			log.Warningf("Error occurred when to get project ID %v", err)
+			return err
+		}
+
+		mfInfo.ProjectID = projectID
+		mfInfo.Refrerence = manifest.References()
+		mfInfo.Digest = desc.Digest.String()
+		blobInfo.ProjectID = projectID
+		blobInfo.Digest = desc.Digest.String()
+		blobInfo.Size = desc.Size
+		blobInfo.ContentType = mediaType
+
+		if err := requireQuota(con, blobInfo); err != nil {
+			return err
+		}
+
+		*req = *(req.WithContext(context.WithValue(req.Context(), util.MFInfokKey, mfInfo)))
+		*req = *(req.WithContext(context.WithValue(req.Context(), util.BBInfokKey, blobInfo)))
+
+		return nil
+	}
+
+	return fmt.Errorf("unsupported content type for manifest: %s", mediaType)
+}
+
+func (sqh *sizeQuotaHandler) handlePutBlobComplete(req *http.Request, blobInfo *util.BlobInfo) error {
+	// the redis connection will be closed in the put response.
+	con, err := util.GetRegRedisCon()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if blobInfo.UUID != "" {
+			_, err := rmBlobUploadUUID(con, blobInfo.UUID)
+			if err != nil {
+				log.Warningf("error occurred when remove UUID for blob, %v", err)
+			}
+		}
+	}()
+
+	dgstStr := req.FormValue("digest")
+	if dgstStr == "" {
+		return errors.New("blob digest missing")
+	}
+	dgst, err := digest.Parse(dgstStr)
+	if err != nil {
+		return errors.New("blob digest parsing failed")
+	}
+
+	blobInfo.Digest = dgst.String()
+	blobInfo.UUID = getUUID(req.URL.Path)
+	size, err := util.GetBlobSize(con, blobInfo.UUID)
+	if err != nil {
+		return err
+	}
+	blobInfo.Size = size
+	if err := requireQuota(con, blobInfo); err != nil {
+		return err
+	}
+	*req = *(req.WithContext(context.WithValue(req.Context(), util.BBInfokKey, blobInfo)))
+	return nil
+
+}
+
+func requireQuota(conn redis.Conn, blobInfo *util.BlobInfo) error {
+	projectID, err := util.GetProjectID(strings.Split(blobInfo.Repository, "/")[0])
+	if err != nil {
+		return err
+	}
+	blobInfo.ProjectID = projectID
+
+	digestLock, err := tryLockBlob(conn, blobInfo)
+	if err != nil {
+		log.Infof("failed to lock digest in redis, %v", err)
+		return err
+	}
+	blobInfo.DigestLock = digestLock
+
+	blobExist, err := dao.HasBlobInProject(blobInfo.ProjectID, blobInfo.Digest)
+	if err != nil {
+		tryFreeBlob(blobInfo)
+		return err
+	}
+	blobInfo.Exist = blobExist
+
+	if !blobExist {
+		quotaRes := &quota.ResourceList{
+			quota.ResourceStorage: blobInfo.Size,
+		}
+		err = util.TryRequireQuota(blobInfo.ProjectID, quotaRes)
+		if err != nil {
+			log.Infof("project id, %d, size %d", blobInfo.ProjectID, blobInfo.Size)
+			tryFreeBlob(blobInfo)
+			log.Errorf("cannot get quota for the blob %v", err)
+			return err
+		}
+		blobInfo.Quota = quotaRes
+	}
+
+	return nil
+}
+
+// tryLockBlob locks blob with redis ...
+func tryLockBlob(conn redis.Conn, blobInfo *util.BlobInfo) (*common_redis.Mutex, error) {
+	digestLock := common_redis.New(conn, "Quota::blob-lock::"+blobInfo.Repository+":"+blobInfo.Digest, common_util.GenerateRandomString())
```

Comment:
Should lock the project+digest rather than repository+digest?

Reply:
done, thanks.